### PR TITLE
Cadence 1.0 Multi-Staging

### LINKED ISSUE #1559: Add ability to stage all contracts
### Problem

Developers have expressed a desire to stage all of their contracts using one command, in order, instead of staging individually.  This helps prevent validation errors, tedious ordering, running multiple commands, and potentially forgetting to stage contracts

### Acceptance Criteria

Some command exists to stage all contracts.  (maybe this is default behaviour for `flow migrate stage-contract`, although, it sounds a bit odd in the singular.  Maybe `flow migrate stage-contract --all`, etc.)

### Context

Mentioned and strongly supported during developers office hours.

### PR DESCRIPTION
**Pull request by jribbink**
Closes #1559 

## Description

Adds support for staging multiple contracts at once.

Screenshot 2024-05-09 at 3 58 33 PM

Approach:
 - Encapsulate staging functionality in "Staging Service"
 - Staging service can stage multiple contracts at once
 - Staging validator adapted to allow multiple contracts, returns separate errors per contract
 - Local dependency tree of all local/remote checkers is built, everything is cached (contracts, checkers, etc).  Checking remote contracts is very expensive, so caching is important.  Additionally, it's important that all contracts are validated _together_, because this way staging order is irrelevant (e.g. validation is done using local copies).
______

For contributor use:

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-cli/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: internal/migrate/stage_contract.go [hunk 1]
```diff
@@ -66,94 +73,251 @@ func stageContract(
...
+	count := 0
+	for _, flag := range flags {
+		if flag {
+			count++
+		}
+	}
+	return count
+}
+
+func pluralize(word string, count int) string {
```
Comment (chasefleming) [inline internal/migrate/stage_contract.go:318]: This might be nice to have in the utils

File: internal/migrate/stage_contract.go [hunk 2]
```diff
@@ -20,35 +20,42 @@ package migrate
...
 	Cmd: &cobra.Command{
-		Use:     "stage-contract <CONTRACT_NAME>",
-		Short:   "stage a contract for migration",
-		Example: `flow migrate stage-contract HelloWorld`,
-		Args:    cobra.MinimumNArgs(1),
+		Use:   "stage-contract [contract names...]",
+		Short: "Stage a contract, or many contracts, for migration",
+		Example: `flow migrate stage-contract Foo Bar --network testnet
+flow migrate stage-contract --account my-account --network testnet
+flow migrate stage-contract --all --network testnet`,
```
Comment (chasefleming) [inline internal/migrate/stage_contract.go:57]: If `--all` is not provided maybe it should stage all by default

File: internal/migrate/stage_contract.go [hunk 3]
```diff
@@ -20,35 +20,42 @@ package migrate
...
+	SkipValidation bool     `default:"false" flag:"skip-validation" info:"Do not validate the contract code against staged dependencies"`
 }
 
 var stageContractCommand = &command.Command{
 	Cmd: &cobra.Command{
-		Use:     "stage-contract <CONTRACT_NAME>",
-		Short:   "stage a contract for migration",
-		Example: `flow migrate stage-contract HelloWorld`,
-		Args:    cobra.MinimumNArgs(1),
+		Use:   "stage-contract [contract names...]",
```
Comment (chasefleming) [inline internal/migrate/stage_contract.go:53]: Is it too late to change this to just `stage`?

File: internal/migrate/stage_contract.go [hunk 4]
```diff
@@ -66,94 +73,251 @@ func stageContract(
...
+		infoMessage.WriteString("\nThese contracts depend on the following contracts which have not been staged yet:\n")
+		missingDependencies := validatorError.MissingDependencies()
+		for _, depLocation := range missingDependencies {
+			infoMessage.WriteString(fmt.Sprintf("  - %s\n", depLocation))
+		}
+
+		infoMessage.WriteString("\nYou may still stage your contract, however it will be unable to be migrated until the missing contracts are staged by their respective owners.  It is important to monitor the status of your contract using the `flow migrate is-validated` command\n")
+		logger.Error(infoMessage.String())
+
+		continuePrompt := promptui.Select{
```
Comment (chasefleming) [inline internal/migrate/stage_contract.go:121]: There is a `GenericBoolPrompt` we should probably start using now. Although I wish we could cherry pick the new prompt changes from the other branch and convert this to using the new library, but I don't want to add more work here. Should be easy enough to change later. Might want to use the `GenericBoolPrompt` though.

File: internal/migrate/stage_contract.go [hunk 5]
```diff
@@ -66,94 +73,251 @@ func stageContract(
...
+) (*stagingResults, error) {
+	contracts, err := state.DeploymentContractsByNetwork(flow.Network())
 	if err != nil {
-		return nil, fmt.Errorf("failed to get cadence string from contract name: %w", err)
+		return nil, err
 	}
 
-	cCode, err := cadence.NewString(string(replacedCode))
+	filteredContracts := make([]*project.Contract, 0)
+	for _, name := range contractNames {
```
Comment (chasefleming) [inline internal/migrate/stage_contract.go:164]: This happens a lot in the codebase. Wonder if this also might make a nice util.

File: internal/migrate/stage_contract.go [hunk 6]
```diff
@@ -66,94 +73,251 @@ func stageContract(
 		return nil, err
 	}
 
-	contractName := args[0]
-	contract, err := state.Contracts().ByName(contractName)
+	// Validate command arguments
+	optionCount := boolCount(stageContractflags.All, len(stageContractflags.Accounts) > 0, len(args) > 0)
+	if optionCount > 1 {
```
Comment (chasefleming) [inline internal/migrate/stage_contract.go:78]: We may also want a max limit. Some arbitrary high enough number not to overwhelm things.

**Review by chasefleming — CHANGES_REQUESTED** [review]

File: internal/migrate/stage_contract.go (hunk 3 above)
Comment (jribbink) [reply internal/migrate/stage_contract.go:53]: I think so, I don't think it would be wise to break functionality right now.

File: internal/migrate/staging_service.go [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+		for location, err := range validatorError.errors {
+			results[location] = stagingResult{
+				err:          err,
+				wasValidated: true,
+			}
+		}
+	}
+
+	// Now, handle contracts that failed validation
+	newResults := s.maybeStageInvalidContracts(ctx, contracts, validatorError)
```
Comment (jribbink) [inline internal/migrate/staging_service.go:120]: make this more readable/where are contracts coming from

File: internal/migrate/staging_service.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+		}
+
+		// Replace imports in the contract
+		script, err := s.flow.ReplaceImportsInScript(context.Background(), rawScript)
+		if err != nil {
+			return nil, fmt.Errorf("failed to replace imports in contract %s: %w", contract.Name, err)
+		}
+
+		// We need the real name of the contract, not the name in flow.json
+		program, err := project.NewProgram(script.Code, script.Args, script.Location)
```
Comment (jribbink) [inline internal/migrate/staging_service.go:271]: can you call ConvertToAddressImports

File: internal/migrate/staging_validator.go [hunk 9]
```diff
@@ -81,44 +100,192 @@ func (e *missingDependenciesError) Error() string {
...
+			if err != nil {
+				errsByAddress[contract.DeployLocation] = err
+			}
+		}
+		return &stagingValidatorError{errors: errsByAddress}
+	}
+	return nil
+}
+
+func (v *stagingValidatorImpl) parseAndCheckAllStaged() map[common.StringLocation]error {
```
Comment (jribbink) [inline internal/migrate/staging_validator.go:248]: checkAllStaged

**Comment by bjartek** [comment]
Just tried to run this on find repo since I have 45 things to stage there and i got
``` 
flow migrate stage-contract --all -n testnet
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
                           ⚠ Upgrade to Cadence 1.0
     The Crescendo network upgrade, including Cadence 1.0, is coming soon.
     You may need to update your existing contracts to support this change.
                     Please visit our migration guide here:
             https://cadence-lang.org/docs/cadence_migration_guide
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
panic: interface conversion: common.Location is common.IdentifierLocation, not common.AddressLocation

goroutine 1 [running]:
github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).parseAndCheckAllStaged.func1({0x104c13e20, 0x140025977a0})
	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:266 +0x128
github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).forEachDependency.func1({0x104c13de0, 0x140024d97e0})
	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:701 +0xfc
github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).forEachDependency.func1({0x104c13de0, 0x14002ac2720})
	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:702 +0x114
github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).forEachDependency(0x14000d477c0, {{{0x140014d0400, 0xb}, {0x3e, 0x5b, 0x4c, 0x62, 0x70, 0x64, 0x62, ...}}, ...}, ...)
	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:706 +0xd0
github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).parseAndCheckAllStaged(0x14000d477c0)
	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:264 +0x1e0
github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).Validate(0x14000d477c0, {0x14000238608, 0x2e, 0x1?})
	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:178 +0x1a8
github.com/onflow/flow-cli/internal/migrate.(*stagingServiceImpl).validateAndStageContracts(0x14000d47800, {0x104c0eb60, 0x1061543e0}, {0x14000238608, 0x2e, 0x4b})
	github.com/onflow/flow-cli/internal/migrate/staging_service.go:101 +0x11c
github.com/onflow/flow-cli/internal/migrate.(*stagingServiceImpl).StageContracts(0x14000d47800, {0x104c0eb60, 0x1061543e0}, {0x14000c31a00?, 0x2e, 0x0?})
	github.com/onflow/flow-cli/internal/migrate/staging_service.go:92 +0x80
github.com/onflow/flow-cli/internal/migrate.stageAll({0x104c000f0, 0x14000d47800}, 0x14000fbcf30, {0x104c359d8?, 0x14000d171a0?})
	github.com/onflow/flow-cli/internal/migrate/stage_contract.go:135 +0xbc
github.com/onflow/flow-cli/internal/migrate.stageContract({0x14000d4b410, 0x0, 0x3}, {{0x0, 0x0}, {0x103ebd395, 0x4}, {0x0, 0x0}, {0x0, ...}, ...}, ...)
	github.com/onflow/flow-cli/internal/migrate/stage_contract.go:92 +0x3cc
github.com/onflow/flow-cli/internal/command.Command.AddToParent.func1(0x14000d51100?, {0x14000d4b410, 0x0, 0x3})
	github.com/onflow/flow-cli/internal/command/command.go:142 +0x4cc
github.com/spf13/cobra.(*Command).execute(0x105fbd4a0, {0x140001b8150, 0x3, 0x3})
	github.com/spf13/cobra@v1.8.0/command.go:987 +0x828
github.com/spf13/cobra.(*Command).ExecuteC(0x1400055d808)
	github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.8.0/command.go:1039
main.main()
	github.com/onflow/flow-cli/cmd/flow/main.go:168 +0xd9c
	```

**Comment by jribbink** [comment]
> Just tried to run this on find repo since I have 45 things to stage there and i got
> 
> ```
> flow migrate stage-contract --all -n testnet
> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
>                            ⚠ Upgrade to Cadence 1.0
>      The Crescendo network upgrade, including Cadence 1.0, is coming soon.
>      You may need to update your existing contracts to support this change.
>                      Please visit our migration guide here:
>              https://cadence-lang.org/docs/cadence_migration_guide
> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
> panic: interface conversion: common.Location is common.IdentifierLocation, not common.AddressLocation
> 
> goroutine 1 [running]:
> github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).parseAndCheckAllStaged.func1({0x104c13e20, 0x140025977a0})
> 	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:266 +0x128
> github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).forEachDependency.func1({0x104c13de0, 0x140024d97e0})
> 	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:701 +0xfc
> github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).forEachDependency.func1({0x104c13de0, 0x14002ac2720})
> 	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:702 +0x114
> github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).forEachDependency(0x14000d477c0, {{{0x140014d0400, 0xb}, {0x3e, 0x5b, 0x4c, 0x62, 0x70, 0x64, 0x62, ...}}, ...}, ...)
> 	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:706 +0xd0
> github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).parseAndCheckAllStaged(0x14000d477c0)
> 	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:264 +0x1e0
> github.com/onflow/flow-cli/internal/migrate.(*stagingValidatorImpl).Validate(0x14000d477c0, {0x14000238608, 0x2e, 0x1?})
> 	github.com/onflow/flow-cli/internal/migrate/staging_validator.go:178 +0x1a8
> github.com/onflow/flow-cli/internal/migrate.(*stagingServiceImpl).validateAndStageContracts(0x14000d47800, {0x104c0eb60, 0x1061543e0}, {0x14000238608, 0x2e, 0x4b})
> 	github.com/onflow/flow-cli/internal/migrate/staging_service.go:101 +0x11c
> github.com/onflow/flow-cli/internal/migrate.(*stagingServiceImpl).StageContracts(0x14000d47800, {0x104c0eb60, 0x1061543e0}, {0x14000c31a00?, 0x2e, 0x0?})
> 	github.com/onflow/flow-cli/internal/migrate/staging_service.go:92 +0x80
> github.com/onflow/flow-cli/internal/migrate.stageAll({0x104c000f0, 0x14000d47800}, 0x14000fbcf30, {0x104c359d8?, 0x14000d171a0?})
> 	github.com/onflow/flow-cli/internal/migrate/stage_contract.go:135 +0xbc
> github.com/onflow/flow-cli/internal/migrate.stageContract({0x14000d4b410, 0x0, 0x3}, {{0x0, 0x0}, {0x103ebd395, 0x4}, {0x0, 0x0}, {0x0, ...}, ...}, ...)
> 	github.com/onflow/flow-cli/internal/migrate/stage_contract.go:92 +0x3cc
> github.com/onflow/flow-cli/internal/command.Command.AddToParent.func1(0x14000d51100?, {0x14000d4b410, 0x0, 0x3})
> 	github.com/onflow/flow-cli/internal/command/command.go:142 +0x4cc
> github.com/spf13/cobra.(*Command).execute(0x105fbd4a0, {0x140001b8150, 0x3, 0x3})
> 	github.com/spf13/cobra@v1.8.0/command.go:987 +0x828
> github.com/spf13/cobra.(*Command).ExecuteC(0x1400055d808)
> 	github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
> github.com/spf13/cobra.(*Command).Execute(...)
> 	github.com/spf13/cobra@v1.8.0/command.go:1039
> main.main()
> 	github.com/onflow/flow-cli/cmd/flow/main.go:168 +0xd9c
> ```

Awesome, thank you so much for doing this.  Looks like an easy fix, glad we could catch it before it became an issue 🙏

**Comment by bjartek** [comment]
this is the output now

```
 flow-ms migrate stage-contract --all -n testnet
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
                           ⚠ Upgrade to Cadence 1.0
     The Crescendo network upgrade, including Cadence 1.0, is coming soon.
     You may need to update your existing contracts to support this change.
                     Please visit our migration guide here:
             https://cadence-lang.org/docs/cadence_migration_guide
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
❌ Preliminary validation could not be performed on the following contracts:
  - 35717efbbce11c74.FindMarketDirectOfferEscrow
  - 3e5b4c627064625d.PartyFavorz
  - 35717efbbce11c74.FindLeaseMarketAuctionSoft
  - 35717efbbce11c74.FindMarketAuctionSoft
  - 35717efbbce11c74.FindLeaseMarketSale
  - 35717efbbce11c74.Dandy
  - 35717efbbce11c74.FindForgeOrder
  - 35717efbbce11c74.FindFurnace
  - 35717efbbce11c74.FindMarketDirectOfferSoft
  - 35717efbbce11c74.FindForge
  - 35717efbbce11c74.FindMarketSale
  - 35717efbbce11c74.FindLeaseMarketDirectOfferSoft
  - 35717efbbce11c74.FindThoughts
  - 3e5b4c627064625d.NFGv3
  - 35717efbbce11c74.FindLeaseMarket
  - 35717efbbce11c74.FindVerifier
  - 35717efbbce11c74.FindMarketAuctionEscrow
  - 35717efbbce11c74.FindLostAndFoundWrapper
  - 35717efbbce11c74.FIND
  - 35717efbbce11c74.FindMarketAdmin
  - 35717efbbce11c74.Admin
  - 35717efbbce11c74.FindAirdropper
  - 35717efbbce11c74.NameVoucher
  - 3e5b4c627064625d.Flomies
  - 35717efbbce11c74.FindPack
  - 3e5b4c627064625d.GeneratedExperiences

These contracts depend on the following contracts which have not been staged yet:
  - a983fecbed621163.FiatToken
  - 0afe396ebc8eee65.FLOAT
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - 0afe396ebc8eee65.FLOAT
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - 0afe396ebc8eee65.FLOAT
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - 0afe396ebc8eee65.FLOAT
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - 0afe396ebc8eee65.FLOAT
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - a983fecbed621163.FiatToken
  - 0afe396ebc8eee65.FLOAT

You may still stage your contract, however it will be unable to be migrated until the missing contracts are staged by their respective owners.  It is important to monitor the status of your contract using the `flow migrate is-validated` command


```

**Comment by jribbink** [comment]
> this is the output now
> 
> ```
>  flow-ms migrate stage-contract --all -n testnet
> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
>                            ⚠ Upgrade to Cadence 1.0
>      The Crescendo network upgrade, including Cadence 1.0, is coming soon.
>      You may need to update your existing contracts to support this change.
>                      Please visit our migration guide here:
>              https://cadence-lang.org/docs/cadence_migration_guide
> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
> ❌ Preliminary validation could not be performed on the following contracts:
>   - 35717efbbce11c74.FindMarketDirectOfferEscrow
>   - 3e5b4c627064625d.PartyFavorz
>   - 35717efbbce11c74.FindLeaseMarketAuctionSoft
>   - 35717efbbce11c74.FindMarketAuctionSoft
>   - 35717efbbce11c74.FindLeaseMarketSale
>   - 35717efbbce11c74.Dandy
>   - 35717efbbce11c74.FindForgeOrder
>   - 35717efbbce11c74.FindFurnace
>   - 35717efbbce11c74.FindMarketDirectOfferSoft
>   - 35717efbbce11c74.FindForge
>   - 35717efbbce11c74.FindMarketSale
>   - 35717efbbce11c74.FindLeaseMarketDirectOfferSoft
>   - 35717efbbce11c74.FindThoughts
>   - 3e5b4c627064625d.NFGv3
>   - 35717efbbce11c74.FindLeaseMarket
>   - 35717efbbce11c74.FindVerifier
>   - 35717efbbce11c74.FindMarketAuctionEscrow
>   - 35717efbbce11c74.FindLostAndFoundWrapper
>   - 35717efbbce11c74.FIND
>   - 35717efbbce11c74.FindMarketAdmin
>   - 35717efbbce11c74.Admin
>   - 35717efbbce11c74.FindAirdropper
>   - 35717efbbce11c74.NameVoucher
>   - 3e5b4c627064625d.Flomies
>   - 35717efbbce11c74.FindPack
>   - 3e5b4c627064625d.GeneratedExperiences
> 
> These contracts depend on the following contracts which have not been staged yet:
>   - a983fecbed621163.FiatToken
>   - 0afe396ebc8eee65.FLOAT
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - 0afe396ebc8eee65.FLOAT
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - 0afe396ebc8eee65.FLOAT
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - 0afe396ebc8eee65.FLOAT
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - 0afe396ebc8eee65.FLOAT
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - a983fecbed621163.FiatToken
>   - 0afe396ebc8eee65.FLOAT
> 
> You may still stage your contract, however it will be unable to be migrated until the missing contracts are staged by their respective owners.  It is important to monitor the status of your contract using the `flow migrate is-validated` command
> ```
> 
> should we sort and unique the dependent contracts maybe?

Yeah 💯

**Comment by bjartek** [comment]
It did work though, but it stages each contract in a single transaction. Each tx takes about 300 gas so you should be able to do loads of them in one go if you want to.

**Comment by jribbink** [comment]
> It did work though, but it stages each contract in a single transaction. Each tx takes about 300 gas so you should be able to do loads of them in one go if you want to.
> 
> The command took 5 minutes to run for me :p

Awesome!!  Super appreciative that you could run this, it really helps because my manual + unit tests can only go so far... lots of edge cases in this one.

Yeah, doing this in separate transactions is somewhat intentional.  I didn't want to run into any headaches related to computation limits, so it seemed more reliable to just stick to individual TX than try to estimate computation & potentially introduce bugs (and since this is what `flow project deploy` does as well anyways).

I know waiting 5 minutes for the command to run isn't a great experience, but hopefully I can really shave that down by not sending unnecessary TX.

**Comment by bjartek** [comment]
Fair enough. I dont expect to run it that many times

File: internal/migrate/staging_service.go (hunk 8 above)
Comment (jribbink) [reply internal/migrate/staging_service.go:257]: This is the reverse of what we are looking for

File: internal/migrate/staging_service.go [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+	stagedContracts := make([]stagedContractUpdate, 0)
+	for _, contract := range contracts {
+		rawScript := flowkit.Script{
+			Code:     contract.Code(),
+			Location: contract.Location(),
+			Args:     contract.Args,
+		}
+
+		// Replace imports in the contract
+		script, err := s.flow.ReplaceImportsInScript(context.Background(), rawScript)
```
Comment (chasefleming) [inline internal/migrate/staging_service.go:251]: Did you try the program method for replacing imports?

Comment (jribbink) [reply internal/migrate/staging_service.go:251]: Yeah, I looked at that one.  It looks like it's for the other way around (e.g. `AddressLocation` -> `StringLocation`, not vice versa)

File: internal/migrate/get_staged_code.go [hunk 11]
```diff
@@ -64,24 +65,56 @@ func getStagedCode(
...
+		return nil, err
+	}
+
+	// If the contract is not staged, return nil
+	if code == nil {
+		return scripts.NewScriptResult(cadence.NewOptional(nil)), nil
 	}
 
-	caddr := cadence.NewAddress(addr)
+	return scripts.NewScriptResult(cadence.NewOptional(cadence.String(code))), nil
```
Comment (jribbink) [inline internal/migrate/get_staged_code.go:79]: Note: this has changed because I tried to extract some of this common functionality.  The result is converted back into Cadence values just so that we can retain the same output (e.g. `scripts.NewScriptResult`).  It's probably not optimal, and probably makes more sense to have a custom output for this command IMO, but out of scope for this PR.

**Review by chasefleming — APPROVED** [review]
